web/interceptor/access_control.go

```diff
@@ -0,0 +1,206 @@
+		for _, role := range roles {
+			logger.Debugf("Checking role: %v", role) // tmp
+			switch role {
+			case none:
+				return handler(ctx, request)
+			case user:
+				if req.IDFor("user") == claims.UserID {
+					// Make sure the user is not updating own admin status.
+					if method == "UpdateUser" {
+						if req.(*qf.User).GetIsAdmin() != claims.Admin {
+							logger.Errorf("Access control: user %d attempted to change admin status from %v to %v",
+								claims.UserID, claims.Admin, req.(*qf.User).GetIsAdmin())
+						}
+					}
+					return handler(ctx, request)
+				}
+			case student:
+				// GetSubmissions is used to fetch individual and group submissions.
+				// For individual submissions needs an extra check for user ID in request.
+				if method == "GetSubmissions" && req.IDFor("group") == 0 {
+					if claims.UserID != req.IDFor("user") {
+						logger.Errorf("AccessControl: ID mismatch for %s in claims (%s) and request (%s)",
+							method, claims.UserID, req.IDFor("user"))
+						return nil, ErrAccessDenied
+					}
+				}
+				courseID := req.IDFor("course")
+				if hasCourseStatus(claims, courseID, qf.Enrollment_STUDENT) {
+					return handler(ctx, request)
+				}
+			case group:
+				// Request for CreateGroup will not have ID yet, need to ckeck
+				// if the user is in the group unless teacher.
+				if method == "CreateGroup" &&
+					!(hasCourseStatus(claims, req.IDFor("course"), qf.Enrollment_TEACHER) ||
+						req.(*qf.Group).Contains(&qf.User{ID: claims.UserID})) {
+					logger.Errorf("AccessControl: user %d creates a new group while not teacher or group member")
+					return nil, ErrAccessDenied
+				}
+				//return nil, fmt.Errorf("Group ID: %d, im claims: %v", req.IDFor("group"), claims.Groups)
+				groupID := req.IDFor("group")
+				for _, group := range claims.Groups {
+					if group == groupID {
+						return handler(ctx, request)
+					}
+				}
+			case teacher:
+				courseID := req.IDFor("course")
+				if hasCourseStatus(claims, courseID, qf.Enrollment_TEACHER) {
+					return handler(ctx, request)
+				}
+			case courseAdmin:
+				if claims.Admin {
+					if method == "GetUserByCourse" {
+						if err := isCourseTeacher(tm.Database(), request.(*qf.CourseUserRequest), claims.Courses); err != nil {
+							logger.Errorf("AccessControl: %v", err)
+							return nil, ErrAccessDenied
+						}
+						return handler(ctx, request)
+					}
+					courseID := req.IDFor("course")
+					if hasCourseStatus(claims, courseID, qf.Enrollment_TEACHER) {
+						return handler(ctx, request)
+					}
+				}
+			case admin:
+				if claims.Admin {
+					return handler(ctx, request)
+				}
+			}
+		}
+		logger.Errorf("%f failed for user %d: insufficient user privileges. Required roles: %v", method, claims.UserID, roles)
+		return nil, ErrAccessDenied
+	}
+}
+
+// hasCourseStatus checks if user is enrolled in a course with a specific status.
+func hasCourseStatus(claims *auth.Claims, courseID uint64, status qf.Enrollment_UserStatus) bool {
```

Comment:
This func should be made into a method on `auth.Claims` and the fields of `Claims` should be unexported unless they need to be exposed. 

The following code:
```go
if ok && currentStatus == status {
	return true
}
return false
```
Is equivalent to:
```go
return ok && currentStatus == status
```
But that said, maybe it would be enough to just do:
```go
func (claims Claims) HasCourseStatus(courseID uint64, status qf.Enrollment_UserStatus) bool {
	return claims.courses[courseID] == status
```
You might want to add a few tests, just to be sure.

Reply:
Perhaps this would be a better version:
```go
func (claims Claims) HasCourseStatus(req requestID, status qf.Enrollment_UserStatus) bool {
	courseID := req.IDFor("course")
	return claims.Courses[courseID] == status
}
```